File: backend-node/models/quality.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const DefectType = sequelize.define('DefectType', {
    type_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    code: {
      type: DataTypes.STRING(50),
      unique: true,
      allowNull: false,
    },
    name: {
      type: DataTypes.STRING(200),
      allowNull: false,
    },
    category: {
      type: DataTypes.STRING(100),
    },
    description: {
      type: DataTypes.TEXT,
    },
    threshold_percentage: {
      type: DataTypes.DECIMAL(5, 2),
    },
    severity: {
      type: DataTypes.ENUM('minor', 'major', 'critical'),
      defaultValue: 'minor',
    },
    is_active: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
    }
  }, {
    tableName: 'defect_types',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: false
  });

  const QCInspection = sequelize.define('QCInspection', {
    inspection_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    inspection_number: {
      type: DataTypes.STRING(50),
      unique: true,
      allowNull: false,
    },
    wo_id: {
      type: DataTypes.INTEGER,
    },
    po_id: {
      type: DataTypes.INTEGER,
    },
    batch_number: {
      type: DataTypes.STRING(50),
      allowNull: false,
    },
    inspection_date: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW,
    },
    inspector_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    quantity_inspected: {
      type: DataTypes.DECIMAL(15, 3),
      allowNull: false,
    },
    quantity_accepted: {
      type: DataTypes.DECIMAL(15, 3),
    },
    quantity_rejected: {
      type: DataTypes.DECIMAL(15, 3),
    },
    result: {
      type: DataTypes.ENUM('pending', 'passed', 'failed', 'conditional'),
      defaultValue: 'pending',
    },
    remarks: {
      type: DataTypes.TEXT,
    },
    approved_by: {
      type: DataTypes.INTEGER,
    },
    approved_at: {
      type: DataTypes.DATE,
    }
  }, {
    tableName: 'qc_inspections',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  const DefectLog = sequelize.define('DefectLog', {
    defect_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    inspection_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    defect_type_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    quantity: {
      type: DataTypes.DECIMAL(15, 3),
      allowNull: false,
    },
    severity: {
      type: DataTypes.ENUM('minor', 'major', 'critical'),
      allowNull: false,
    },
    location: {
      type: DataTypes.STRING(200),
    },
    description: {
      type: DataTypes.TEXT,
    },
    root_cause: {
      type: DataTypes.TEXT,
    },
    corrective_action: {
      type: DataTypes.TEXT,
    },
    image_url: {
      type: DataTypes.STRING(500),
    }
  }, {
    tableName: 'defect_logs',
    timestamps: true,
    createdAt: 'logged_at',
    updatedAt: false
  });

  const BatchApproval = sequelize.define('BatchApproval', {
    approval_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    batch_number: {
      type: DataTypes.STRING(50),
      allowNull: false,
    },
    inspection_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    status: {
      type: DataTypes.STRING(20),
      defaultValue: 'pending', // "pending", "approved", "rejected", "rework"
    },
    approved_by: {
      type: DataTypes.INTEGER,
    },
    approved_at: {
      type: DataTypes.DATE,
    },
    rejection_reason: {
      type: DataTypes.TEXT,
    },
    rework_instructions: {
      type: DataTypes.TEXT,
    },
    notes: {
      type: DataTypes.TEXT,
    }
  }, {
    tableName: 'batch_approvals',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  return { DefectType, QCInspection, DefectLog, BatchApproval };
};

File: backend/app/models/quality.py
"""
Quality Module Models
QC Inspections, Defects, Batch Approvals
"""

from sqlalchemy import Column, Integer, String, Numeric, DateTime, ForeignKey, Text, Enum as SQLEnum, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from app.core.database import Base

class InspectionResult(str, enum.Enum):
    PENDING = "pending"
    PASSED = "passed"
    FAILED = "failed"
    CONDITIONAL = "conditional"

class DefectSeverity(str, enum.Enum):
    MINOR = "minor"
    MAJOR = "major"
    CRITICAL = "critical"

class DefectType(Base):
    """Types of defects"""
    __tablename__ = "defect_types"
    
    type_id = Column(Integer, primary_key=True, index=True)
    code = Column(String(50), unique=True, nullable=False)
    name = Column(String(200), nullable=False)
    category = Column(String(100))
    description = Column(Text)
    threshold_percentage = Column(Numeric(5, 2))  # Acceptable limit
    severity = Column(SQLEnum(DefectSeverity), default=DefectSeverity.MINOR)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    defect_logs = relationship("DefectLog", back_populates="defect_type")

class QCInspection(Base):
    """Quality control inspections"""
    __tablename__ = "qc_inspections"
    
    inspection_id = Column(Integer, primary_key=True, index=True)
    inspection_number = Column(String(50), unique=True, nullable=False, index=True)
    wo_id = Column(Integer, ForeignKey("work_orders.wo_id"), index=True)
    po_id = Column(Integer, ForeignKey("purchase_orders.po_id"), index=True)
    batch_number = Column(String(50), nullable=False, index=True)
    inspection_date = Column(DateTime(timezone=True), server_default=func.now())
    inspector_id = Column(Integer, ForeignKey("users.user_id"), nullable=False)
    quantity_inspected = Column(Numeric(15, 3), nullable=False)
    quantity_accepted = Column(Numeric(15, 3))
    quantity_rejected = Column(Numeric(15, 3))
    result = Column(SQLEnum(InspectionResult), default=InspectionResult.PENDING, index=True)
    remarks = Column(Text)
    approved_by = Column(Integer, ForeignKey("users.user_id"))
    approved_at = Column(DateTime(timezone=True))
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    work_order = relationship("WorkOrder", back_populates="qc_inspections")
    defect_logs = relationship("DefectLog", back_populates="inspection")

class DefectLog(Base):
    """Defect details logged during inspection"""
    __tablename__ = "defect_logs"
    
    defect_id = Column(Integer, primary_key=True, index=True)
    inspection_id = Column(Integer, ForeignKey("qc_inspections.inspection_id"), nullable=False, index=True)
    defect_type_id = Column(Integer, ForeignKey("defect_types.type_id"), nullable=False)
    quantity = Column(Numeric(15, 3), nullable=False)
    severity = Column(SQLEnum(DefectSeverity), nullable=False)
    location = Column(String(200))  # Where defect was found
    description = Column(Text)
    root_cause = Column(Text)
    corrective_action = Column(Text)
    image_url = Column(String(500))
    logged_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    inspection = relationship("QCInspection", back_populates="defect_logs")
    defect_type = relationship("DefectType", back_populates="defect_logs")

class BatchApproval(Base):
    """Batch approval workflow"""
    __tablename__ = "batch_approvals"
    
    approval_id = Column(Integer, primary_key=True, index=True)
    batch_number = Column(String(50), nullable=False, index=True)
    inspection_id = Column(Integer, ForeignKey("qc_inspections.inspection_id"), nullable=False)
    status = Column(String(20), default="pending")  # "pending", "approved", "rejected", "rework"
    approved_by = Column(Integer, ForeignKey("users.user_id"))
    approved_at = Column(DateTime(timezone=True))
    rejection_reason = Column(Text)
    rework_instructions = Column(Text)
    notes = Column(Text)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
